# Coding Exercises (Part 8)

## Pandas: Advanced Topics

### Exercise 8: Creating and Removing Rows and Columns

In [1]:
#run the cell!
import pandas as pd
pd.options.display.float_format = '{:.2f}'.format

In [2]:
#run the cell!
listings = pd.read_csv("listings_clean.csv")

In [3]:
#run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price     Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     AA     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   AABA   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3    AAC     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year            Sector  \
0      1999     Capital Goods   
1      2016  Basic Industries   
2      1969        Technology   
3      2014       Health Care   
4      1969    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [4]:
# run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6852 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   IPO_Year    6852 non-null   int64  
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 428.4+ KB


76. __Drop__ the columns __"Symbol"__ and __"Sector"__ from the DataFrame listings! __Save__ the change! __Fill in the gaps!__

In [3]:
listings.drop(columns = ["Symbol", "Sector"], inplace = True)

In [4]:
# run the cell!
listings.head()

Exchange                           Name  Last_Price     Market_Cap  \
0     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year                                          Industry  
0      1999  Biotechnology: Laboratory Analytical Instruments  
1      2016                                          Aluminum  
2      1969                                      EDP Services  
3      2014                              Medical Specialities  
4      1969                     Air Freight/Delivery Services

77. __Drop__ all rows/listings on __"AMEX"__ from the DataFrame listings! Use the __loc__ operator and __overwrite listings__!<br> __Fill in the gaps!__

In [5]:
listings = listings.loc[listings.Exchange!="AMEX"]

In [6]:
# run the cell!
listings.head()

Exchange                           Name  Last_Price     Market_Cap  \
0     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year                                          Industry  
0      1999  Biotechnology: Laboratory Analytical Instruments  
1      2016                                          Aluminum  
2      1969                                      EDP Services  
3      2014                              Medical Specialities  
4      1969                     Air Freight/Delivery Services

In [10]:
# run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6541 entries, 0 to 6851
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Exchange    6541 non-null   object 
 1   Name        6541 non-null   object 
 2   Last_Price  6541 non-null   float64
 3   Market_Cap  5677 non-null   float64
 4   IPO_Year    6541 non-null   int64  
 5   Industry    5085 non-null   object 
dtypes: float64(2), int64(1), object(3)
memory usage: 357.7+ KB


78. __Add__ the new column __MCAP_MUSD__ (Market Cap in Million USD) to the DataFrame listings by dividing Market_Cap by 1 million. __Fill in the gaps__!

In [12]:
listings["MCAP_MUSD"] = listings.Market_Cap/1000000

/var/folders/b8/lmz1tyx9785389sr236542740000gn/T/ipykernel_2153/2102827983.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  listings["MCAP_MUSD"] = listings.Market_Cap/1000000


__Inspect__! What is the __MCAP_MUSD__ for AAC Holdings?

In [13]:
# run the cell!
listings.head()

Exchange                           Name  Last_Price     Market_Cap  \
0     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year                                          Industry  MCAP_MUSD  
0      1999  Biotechnology: Laboratory Analytical Instruments   25934.70  
1      2016                                          Aluminum    5407.81  
2      1969                                      EDP Services   42781.13  
3      2014                              Medical Specialities      53.14  
4      1969                     Air Freight/Delivery Services   15276.87

AAC Holdings has a Market Cap of ... 53.14 MUSD!

79. __Add__ the new column __Shares__ with the total amount of __shares outstanding__ for each listing. __Divide__ __Market_Cap__ by the __Last_Price__! <br>
Instead of using the arithmetic sign "/", use an __appropriate method for division__! 

In [7]:
listings["Shares"] = listings["Market_Cap"].divide(listings["Last_Price"])

In [8]:
# run the cell!
listings.head()

Exchange                           Name  Last_Price     Market_Cap  \
0     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year                                          Industry       Shares  
0      1999  Biotechnology: Laboratory Analytical Instruments 317515869.00  
1      2016                                          Aluminum 185516624.00  
2      1969                                      EDP Services 567464270.00  
3      2014                              Medical Specialities  24602355.00  
4      1969                     Air Freight/Delivery Services 449055548.00

In [17]:
# run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6541 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Exchange    6541 non-null   object 
 1   Name        6541 non-null   object 
 2   Last_Price  6541 non-null   float64
 3   Market_Cap  5677 non-null   float64
 4   IPO_Year    6541 non-null   int64  
 5   Industry    5085 non-null   object 
 6   MCAP_MUSD   5677 non-null   float64
 7   Shares      5583 non-null   float64
dtypes: float64(4), int64(1), object(3)
memory usage: 459.9+ KB


80. __Sort__ listings by the __Shares__ column from __high to low__. Which listing has the __largest number__ of shares outstanding

In [18]:
listings.sort_values("Shares",ascending=False)

Exchange                                Name  Last_Price     Market_Cap  \
3812     NYSE            Lloyds Banking Group Plc        3.28 58377889642.08   
5422     NYSE               Banco Santander, S.A.        4.91 79721578055.22   
25       NYSE                          Ambev S.A.        4.54 71397713602.90   
4169     NYSE  Mitsubishi UFJ Financial Group Inc        5.04 65129451176.16   
3929     NYSE        Mizuho Financial Group, Inc.        3.12 39558969655.92   
...       ...                                 ...         ...            ...   
6819     NYSE           Zions Bancorporation N.A.       27.62            NaN   
6820     NYSE           Zions Bancorporation N.A.       25.55            NaN   
6830   NASDAQ           Zions Bancorporation N.A.       14.74            NaN   
6839   NASDAQ                  Zion Oil & Gas Inc        0.00           0.00   
6843   NASDAQ                   BATS BZX Exchange       10.10            NaN   

      IPO_Year                             Industry  MCAP_MUSD         Shares  
3812      1969                     Commercial Banks   58377.89 17798137086.00  
5422      1969                     Commercial Banks   79721.58 16236573942.00  
25        2013  Beverages (Production/Distribution)   71397.71 15726368635.00  
4169      1969                     Commercial Banks   65129.45 12922510154.00  
3929      1969                          Major Banks   39558.97 12679156941.00  
...        ...                                  ...        ...            ...  
6819      1969                                  NaN        NaN            NaN  
6820      1969                                  NaN        NaN            NaN  
6830      1969                          Major Banks        NaN            NaN  
6839      1969                 Oil & Gas Production       0.00            NaN  
6843      1969                                  NaN        NaN            NaN  

[6541 rows x 8 columns]

Lloyds Banking Group has over 17 billion shares outstanding!

81. Let´s assume we have the listings for NYSE, NASDAQ and AMEX in __three separate datasets__. __Import__ the three datasets into Pandas and __inspect__!

In [9]:
# run the cell!
nyse = pd.read_csv("NYSE.csv")
nasdaq = pd.read_csv("NASDAQ.csv") 
amex = pd.read_csv("AMEX.csv") 

In [10]:
# run the cell!
nyse.head()

Symbol                    Name  LastSale       MarketCap  ADR TSO  IPOyear  \
0    DDD  3D Systems Corporation     10.91   1272949528.47      NaN      NaN   
1    MMM              3M Company    212.11 122153435461.96      NaN      NaN   
2   WBAI         500.com Limited     13.85    588423939.55      NaN  2013.00   
3   WUBA             58.com Inc.     70.11  10391864681.79      NaN  2013.00   
4   EGHT                 8x8 Inc     21.09   2017357580.82      NaN      NaN   

              Sector                                         Industry  
0         Technology          Computer Software: Prepackaged Software  
1        Health Care                       Medical/Dental Instruments  
2  Consumer Services            Services-Misc. Amusement & Recreation  
3         Technology  Computer Software: Programming, Data Processing  
4         Technology                                     EDP Services

In [21]:
# run the cell!
nyse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3106 entries, 0 to 3105
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Symbol     3106 non-null   object 
 1   Name       3106 non-null   object 
 2   LastSale   3015 non-null   float64
 3   MarketCap  3106 non-null   float64
 4   ADR TSO    0 non-null      float64
 5   IPOyear    1416 non-null   float64
 6   Sector     2121 non-null   object 
 7   Industry   2121 non-null   object 
dtypes: float64(4), object(4)
memory usage: 194.2+ KB


In [22]:
# run the cell!
nasdaq.head()

Symbol                                    Name  LastSale     MarketCap  \
0     YI                               111, Inc.      7.10   56690340.50   
1    PIH  1347 Property Insurance Holdings, Inc.      5.20   31266372.80   
2  PIHPP  1347 Property Insurance Holdings, Inc.     24.90   17430000.00   
3   TURN                180 Degree Capital Corp.      1.91   59286575.61   
4   FLWS                 1-800 FLOWERS.COM, Inc.     18.81 1206817397.28   

     ADR TSO  IPOyear             Sector                    Industry  
0 7984555.00  2018.00        Health Care    Medical/Nursing Services  
1        NaN  2014.00            Finance  Property-Casualty Insurers  
2        NaN      NaN            Finance  Property-Casualty Insurers  
3        NaN      NaN            Finance  Finance/Investors Services  
4        NaN  1999.00  Consumer Services      Other Specialty Stores

In [23]:
# run the cell!
nasdaq.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3435 entries, 0 to 3434
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Symbol     3435 non-null   object 
 1   Name       3435 non-null   object 
 2   LastSale   3432 non-null   float64
 3   MarketCap  3435 non-null   float64
 4   ADR TSO    140 non-null    float64
 5   IPOyear    1599 non-null   float64
 6   Sector     2964 non-null   object 
 7   Industry   2964 non-null   object 
dtypes: float64(4), object(4)
memory usage: 214.8+ KB


In [24]:
# run the cell!
amex.head()

Symbol                                               Name  LastSale  \
0   XXII                            22nd Century Group, Inc      1.89   
1    FAX              Aberdeen Asia-Pacific Income Fund Inc      4.14   
2    IAF                 Aberdeen Australia Equity Fund Inc      5.45   
3    AEF  Aberdeen Emerging Markets Equity Income Fund, ...      7.58   
4    FCO                  Aberdeen Global Income Fund, Inc.      8.16   

      MarketCap  ADR TSO  IPOyear                 Sector  \
0  235574500.77      NaN      NaN  Consumer Non-Durables   
1 1042993611.36      NaN  1986.00                    NaN   
2  123945676.70      NaN      NaN                    NaN   
3  453154667.63      NaN      NaN                    NaN   
4   71194278.24      NaN  1992.00                    NaN   

                Industry  
0  Farming/Seeds/Milling  
1                    NaN  
2                    NaN  
3                    NaN  
4                    NaN

In [25]:
# run the cell!
amex.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 311 entries, 0 to 310
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Symbol     311 non-null    object 
 1   Name       311 non-null    object 
 2   LastSale   298 non-null    float64
 3   MarketCap  311 non-null    float64
 4   ADR TSO    0 non-null      float64
 5   IPOyear    90 non-null     float64
 6   Sector     224 non-null    object 
 7   Industry   224 non-null    object 
dtypes: float64(4), object(4)
memory usage: 19.6+ KB


82. __Concatenate__ the three DataFrames into one DataFrame and save the new DataFrame in the variable __all_listings__! <br>
Each row/listing should contain information about the respective __Exchange__ (__NYSE__, __NASDAQ__ or __AMEX__)! __Fill in the gaps__!

In [26]:
all_listings = pd.concat([nyse, nasdaq, amex], keys = ["NYSE", "NASDAQ", "AMEX"], names = ["Exchange"])

In [27]:
# run the cell!
all_listings

Symbol                            Name  LastSale       MarketCap  \
Exchange                                                                        
NYSE     0      DDD          3D Systems Corporation     10.91   1272949528.47   
         1      MMM                      3M Company    212.11 122153435461.96   
         2     WBAI                 500.com Limited     13.85    588423939.55   
         3     WUBA                     58.com Inc.     70.11  10391864681.79   
         4     EGHT                         8x8 Inc     21.09   2017357580.82   
...             ...                             ...       ...             ...   
AMEX     306    WTT   Wireless Telecom Group,  Inc.      1.55     33015390.60   
         307   XTNT    Xtant Medical Holdings, Inc.      3.08     40538226.96   
         308   YUMA               Yuma Energy, Inc.      0.23      5443343.78   
         309   ZDGE                     Zedge, Inc.      1.91     18763769.33   
         310    ZOM  Zomedica Pharmaceuticals Corp.      0.33     35652671.34   

              ADR TSO  IPOyear             Sector  \
Exchange                                            
NYSE     0        NaN      NaN         Technology   
         1        NaN      NaN        Health Care   
         2        NaN  2013.00  Consumer Services   
         3        NaN  2013.00         Technology   
         4        NaN      NaN         Technology   
...               ...      ...                ...   
AMEX     306      NaN      NaN      Capital Goods   
         307      NaN      NaN        Health Care   
         308      NaN      NaN             Energy   
         309      NaN  2016.00         Technology   
         310      NaN  2017.00        Health Care   

                                                       Industry  
Exchange                                                         
NYSE     0              Computer Software: Prepackaged Software  
         1                           Medical/Dental Instruments  
         2                Services-Misc. Amusement & Recreation  
         3      Computer Software: Programming, Data Processing  
         4                                         EDP Services  
...                                                         ...  
AMEX     306                                Electrical Products  
         307  Biotechnology: Biological Products (No Diagnos...  
         308                               Oil & Gas Production  
         309            Computer Software: Prepackaged Software  
         310                              Major Pharmaceuticals  

[6852 rows x 8 columns]

In [28]:
# run the cell to tidy up!
all_listings.reset_index(drop=True,inplace=True)
# reset_index does not return any value

How many listings does __listings_all__ contain in total?

In [29]:
# run and inspect!
all_listings

Symbol                            Name  LastSale       MarketCap  \
0       DDD          3D Systems Corporation     10.91   1272949528.47   
1       MMM                      3M Company    212.11 122153435461.96   
2      WBAI                 500.com Limited     13.85    588423939.55   
3      WUBA                     58.com Inc.     70.11  10391864681.79   
4      EGHT                         8x8 Inc     21.09   2017357580.82   
...     ...                             ...       ...             ...   
6847    WTT   Wireless Telecom Group,  Inc.      1.55     33015390.60   
6848   XTNT    Xtant Medical Holdings, Inc.      3.08     40538226.96   
6849   YUMA               Yuma Energy, Inc.      0.23      5443343.78   
6850   ZDGE                     Zedge, Inc.      1.91     18763769.33   
6851    ZOM  Zomedica Pharmaceuticals Corp.      0.33     35652671.34   

      ADR TSO  IPOyear             Sector  \
0         NaN      NaN         Technology   
1         NaN      NaN        Health Care   
2         NaN  2013.00  Consumer Services   
3         NaN  2013.00         Technology   
4         NaN      NaN         Technology   
...       ...      ...                ...   
6847      NaN      NaN      Capital Goods   
6848      NaN      NaN        Health Care   
6849      NaN      NaN             Energy   
6850      NaN  2016.00         Technology   
6851      NaN  2017.00        Health Care   

                                               Industry  
0               Computer Software: Prepackaged Software  
1                            Medical/Dental Instruments  
2                 Services-Misc. Amusement & Recreation  
3       Computer Software: Programming, Data Processing  
4                                          EDP Services  
...                                                 ...  
6847                                Electrical Products  
6848  Biotechnology: Biological Products (No Diagnos...  
6849                               Oil & Gas Production  
6850            Computer Software: Prepackaged Software  
6851                              Major Pharmaceuticals  

[6852 rows x 8 columns]

In [34]:
# run and inspect!
all_listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Exchange   6852 non-null   object 
 1   Symbol     6852 non-null   object 
 2   Name       6852 non-null   object 
 3   LastSale   6745 non-null   float64
 4   MarketCap  6852 non-null   float64
 5   ADR TSO    140 non-null    float64
 6   IPOyear    3105 non-null   float64
 7   Sector     5309 non-null   object 
 8   Industry   5309 non-null   object 
dtypes: float64(4), object(5)
memory usage: 481.9+ KB


There are in total ... 6852 listings!

# Recap

76. drop columns of dataframe

`df.drop(inplace=True)`

note:drop method does not return any value

79. do calculations with dataframe columns

`df1.Col1.divide(df2.Col2)`

82. append dataframes and indicate data sources


`pd.concat(df_list, keys=key_list, names=name_list)`